include/cuco/detail/static_multimap.inl

```diff
@@ -0,0 +1,1153 @@
+
+      if (g.any(equals)) { found_match = true; }
+
+      thread_num_matches += equals;
+
+      if (g.any(slot_is_empty)) {
+        if ((not found_match) && (g.thread_rank() == 0)) { thread_num_matches++; }
+        break;
+      }
+
+      current_slot = next_slot(current_slot);
+    }
+  } else {
+    while (true) {
+      auto slot_contents = *reinterpret_cast<cuco::pair_type<Key, Value> const*>(current_slot);
+
+      auto const slot_is_empty = (slot_contents.first == this->get_empty_key_sentinel());
+
+      auto const equals = not slot_is_empty and pair_equal(slot_contents, pair);
+
+      thread_num_matches += equals;
+
+      if (g.any(slot_is_empty)) { break; }
+
+      current_slot = next_slot(current_slot);
+    }
+  }
+}
+
+template <typename Key,
+          typename Value,
+          class ProbeSequence,
+          cuda::thread_scope Scope,
+          typename Allocator>
+template <uint32_t buffer_size,
+          bool is_vector_load,
+          bool is_outer,
+          typename CG,
+          typename atomicT,
+          typename OutputIt,
+          typename KeyEqual>
+__device__ std::enable_if_t<is_vector_load, void>
+static_multimap<Key, Value, ProbeSequence, Scope, Allocator>::device_view::retrieve(
+  const unsigned int activemask,
+  CG const& g,
+  Key const& k,
+  uint32_t* warp_counter,
+  value_type* output_buffer,
+  atomicT* num_matches,
+  OutputIt output_begin,
+  KeyEqual key_equal) noexcept
+{
+  const uint32_t warp_lane_id = threadIdx.x % 32;
+  const uint32_t cg_lane_id   = g.thread_rank();
+
+  auto current_slot = initial_slot(g, k);
+
+  bool running = true;
+  if constexpr (is_outer) {
+    bool found_match = false;
+
+    while (__any_sync(activemask, running)) {
+      if (running) {
+        pair<Key, Value> arr[2];
+        if constexpr (sizeof(Key) == 4) {
+          auto const tmp = *reinterpret_cast<uint4 const*>(current_slot);
+          memcpy(&arr[0], &tmp, 2 * sizeof(pair<Key, Value>));
+        } else {
+          auto const tmp = *reinterpret_cast<ulonglong4 const*>(current_slot);
+          memcpy(&arr[0], &tmp, 2 * sizeof(pair<Key, Value>));
+        }
+
+        auto const first_slot_is_empty  = (arr[0].first == this->get_empty_key_sentinel());
+        auto const second_slot_is_empty = (arr[1].first == this->get_empty_key_sentinel());
+        auto const first_equals         = (not first_slot_is_empty and key_equal(arr[0].first, k));
+        auto const second_equals        = (not second_slot_is_empty and key_equal(arr[1].first, k));
+        auto const first_exists         = g.ballot(first_equals);
+        auto const second_exists        = g.ballot(second_equals);
+
+        if (first_exists or second_exists) {
+          found_match = true;
+
+          auto num_first_matches  = __popc(first_exists);
+          auto num_second_matches = __popc(second_exists);
+
+          uint32_t output_idx;
+          if (0 == cg_lane_id) {
```

Comment:
Should this be `warp_lane_id == 0` instead? I guess it's a little unclear why `warp_counter` (which is per-warp?) is getting incremented by all threads of rank 0 for each CG (which is not necessarily = warp?)

Reply:
We've tested the shared memory buffer at 3 different levels: CG, warp, and block. What we discovered is that using a per-CG shared memory buffer together with vector load will make the program "atomic-bound" cause we atomically update the global counter too often. For example, if the per-CG buffer size is 16, we will do atomic add over the global counter for every 16 matches in the ideal case. As for the per-block counter, however, the buffer size is 256 so we do the same atomic operation much less often (one atomic add per 256 matches v.s. one atomic add per 16 matches). The problem of the per-block buffer is that it will degrade performance when multiplicities are low. As a compromise between per-block buffer and per-cg buffer, the per-warp buffer won't do atomic add too often and it can keep relatively good performance for low multiplicities.